File: Sources/Core/TCParser.cpp
/*
 *  TCParser.cpp
 *
 *  Copyright 2010 Avérous Julien-Pierre
 *
 *  This file is part of TorChat.
 *
 *  TorChat is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  TorChat is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with TorChat.  If not, see <http://www.gnu.org/licenses/>.
 *
 */



#include "TCParser.h"

#include "TCTools.h"
#include "TCInfo.h"



/*
** TCParser - Parsing
*/
#pragma mark -
#pragma mark TCParser - Parsing

// == Give a line to to be parsed to the parser ==
void TCParser::parseLine(const std::string &line)
{
	// Unscape protocol special chars
	std::string *l1 = createReplaceAll(line, "\\n", "\n");
	std::string *l2 = createReplaceAll(*l1, "\\/", "\\");
	
	// Eplode the line with spaces
	std::vector<std::string> *exp = createExplode(*l2, " ");
	
	// Parse the array
	_parseCommand(*exp);
	
	// Free memory
	delete l1;
	delete l2;
	delete exp;	
}



/*
** TCParser - Command
*/
#pragma mark -
#pragma mark TCParser - Command

// -- To be overwritten --
void TCParser::doPing(const std::string &address, const std::string &random)
{
	_parserError(tcrec_cmd_ping, "Ping: Not handled");
}

void TCParser::doPong(const std::string &random)
{
	_parserError(tcrec_cmd_pong, "Pong: Not handled");
}

void TCParser::doStatus(const std::string &status)
{
	_parserError(tcrec_cmd_status, "Status: Not handled");
}

void TCParser::doMessage(const std::string &message)
{
	_parserError(tcrec_cmd_message, "Message: Not handled");
}

void TCParser::doVersion(const std::string &version)
{
	_parserError(tcrec_cmd_version, "Version: Not handled");
}

void TCParser::doAddMe()
{
	_parserError(tcrec_cmd_addme, "AddMe:  Not handled");
}

void TCParser::doRemoveMe()
{
	_parserError(tcrec_cmd_removeme, "RemoveMe: Not handled");
}

void TCParser::doFileName(const std::string &uuid, const std::string &fsize, const std::string &bsize, const std::string &filename)
{
	_parserError(tcrec_cmd_filename, "FileName: Not handled");
}

void TCParser::doFileData(const std::string &uuid, const std::string &start, const std::string &hash, const std::string &data)
{
	_parserError(tcrec_cmd_filedata, "FileData: Not handled");
}

void TCParser::doFileDataB64(const std::string &uuid, const std::string &start, const std::string &hash, const std::string &data)
{
	_parserError(tcrec_cmd_filedata_b64, "FileDataB64:  Not handled");
}

void TCParser::doFileDataOk(const std::string &uuid, const std::string &start)
{
	_parserError(tcrec_cmd_filedataok, "FileDataOk: Not handled");
}

void TCParser::doFileDataError(const std::string &uuid, const std::string &start)
{
	_parserError(tcrec_cmd_filedataerror, "FileDataError: Not handled");
}

void TCParser::doFileStopSending(const std::string &uuid)
{
	_parserError(tcrec_cmd_filestopsending, "FileStopSending: Not handled");
}

void TCParser::doFileStopReceiving(const std::string &uuid)
{
	_parserError(tcrec_cmd_filestopreceiving, "FileStopReceiving: Not handled");
}



/*
** TCParser - Error
*/
#pragma mark -
#pragma mark TCParser - Error

void TCParser::parserError(TCInfo *info)
{
	fprintf(stderr, "Unhandled error (%s)\n", info->render().c_str());
}



/*
** TCParser - Parser
*/
#pragma mark -
#pragma mark TCParser - Parser

void TCParser::_parseCommand(std::vector<std::string> &items)
{
	if (items.size() == 0)
        return;
    
    std::string command = items[0];
	
	items.erase(items.begin());
    
    // Dispatch command
    if (command.compare("ping") == 0)
        _parsePing(items);
    else if (command.compare("pong") == 0)
        _parsePong(items);
    else if (command.compare("status") == 0)
        _parseStatus(items);
    else if (command.compare("version") == 0)
        _parseVersion(items);
	else if (command.compare("message") == 0)
        _parseMessage(items);
	else if (command.compare("add_me") == 0)
        _parseAddMe(items);
	else if (command.compare("remove_me") == 0)
        _parseRemoveMe(items);
	else if (command.compare("filename") == 0)
        _parseFileName(items);
	else if (command.compare("filedata") == 0)
        _parseFileData(items);
	else if (command.compare("filedata_b64") == 0)
        _parseFileDataB64(items);
	else if (command.compare("filedata_ok") == 0)
        _parseFileDataOk(items);
	else if (command.compare("filedata_error") == 0)
        _parseFileDataError(items);
	else if (command.compare("file_stop_sending") == 0)
        _parseFileStopSending(items);
	else if (command.compare("file_stop_receiving") == 0)
        _parseFileStopReceiving(items);
    else
		_parserError(tcrec_unknown_command, "Unknown command");
}

void TCParser::_parsePing(const std::vector<std::string> &args)
{
	if (args.size() != 2)
    {
		_parserError(tcrec_cmd_ping, "Bad ping argument");
        return;
    }
	
	doPing(args[0], args[1]);
}

void TCParser::_parsePong(const std::vector<std::string> &args)
{
	if (args.size() != 1)
    {
		_parserError(tcrec_cmd_pong, "Bad pong argument");
        return;
	}
	
	doPong(args[0]);
}

void TCParser::_parseStatus(const std::vector<std::string> &args)
{
	if (args.size() != 1)
    {
		_parserError(tcrec_cmd_status, "Bad status argument");
        return;
	}
	
	doStatus(args[0]);
}

void TCParser::_parseVersion(const std::vector<std::string> &args)
{
	if (args.size() != 1)
    {
		_parserError(tcrec_cmd_version, "Bad version argument");
        return;
	}
	
	doVersion(args[0]);
}

void TCParser::_parseMessage(const std::vector<std::string> &args)
{
	if (args.size() == 0)
    {
		_parserError(tcrec_cmd_message, "Empty message content");
        return;
	}
	
	std::string * msg = createJoin(args, " ");
	
	doMessage(*msg);
	
	delete msg;
}

void TCParser::_parseAddMe(const std::vector<std::string> &args)
{
	doAddMe();
}

void TCParser::_parseRemoveMe(const std::vector<std::string> &args)
{
	doRemoveMe();
}

void TCParser::_parseFileName(const std::vector<std::string> &args)
{
	if (args.size() != 4)
    {
		_parserError(tcrec_cmd_filename, "Bad filename argument");
        return;
	}
	
	doFileName(args[0], args[1], args[2], args[3]);
}

void TCParser::_parseFileData(const std::vector<std::string> &args)
{
	if (args.size() < 4)
    {
		_parserError(tcrec_cmd_filedata, "Bad filedata argument");
		
        return;
	}
	
	std::string *data = createJoin(args, 3, " ");
	
	if (data)
	{
		doFileData(args[0], args[1], args[2], *data);
		
		delete data;
	}
	else
		doFileData(args[0], args[1], args[2], "");
}

void TCParser::_parseFileDataB64(const std::vector<std::string> &args)
{
	if (args.size() != 4)
    {
		_parserError(tcrec_cmd_filedata, "Bad filedata_b64 argument");
        return;
	}
	
	doFileDataB64(args[0], args[1], args[2], args[3]);
}

void TCParser::_parseFileDataOk(const std::vector<std::string> &args)
{
	if (args.size() != 2)
    {
		_parserError(tcrec_cmd_filedataok, "Bad filedataok argument");
        return;
	}
	
	doFileDataOk(args[0], args[1]);
}

void TCParser::_parseFileDataError(const std::vector<std::string> &args)
{
	if (args.size() != 2)
    {
		_parserError(tcrec_cmd_filedataerror, "Bad filedataerror argument");
        return;
	}

	doFileDataError(args[0], args[1]);
}

void TCParser::_parseFileStopSending(const std::vector<std::string> &args)
{
	if (args.size() != 1)
    {
		_parserError(tcrec_cmd_filestopsending, "Bad filestopsending argument");
        return;
	}
	
	doFileStopSending(args[0]);
}

void TCParser::_parseFileStopReceiving(const std::vector<std::string> &args)
{
	if (args.size() != 1)
    {
		_parserError(tcrec_cmd_filestopreceiving, "Bad filestopreceiving argument");
        return;
	}

	doFileStopReceiving(args[0]);
}


void TCParser::_parserError(tcrec_error error, const char *info)
{
	TCInfo *err = new TCInfo(tcinfo_error, error, info);
	
	parserError(err);
	
	err->release();
}

File: TorChat/Sources/Core/TCParser.m
/*
 *  TCParser.cpp
 *
 *  Copyright 2014 Avérous Julien-Pierre
 *
 *  This file is part of TorChat.
 *
 *  TorChat is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  TorChat is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with TorChat.  If not, see <http://www.gnu.org/licenses/>.
 *
 */



#import "TCParser.h"

#import "TCTools.h"

#import "NSData+TCTools.h"
#import "NSArray+TCTools.h"



/*
** TCParser - Private
*/
#pragma mark - TCParser - Private

@interface TCParser ()
{
	__weak id <TCParserCommand> _receiver;
}

@end



/*
** TCParser
*/
#pragma mark - TCParser

@implementation TCParser


/*
** TCParser - Instance
*/
#pragma mark - TCParser - Instance

- (id)initWithParsingResult:(id <TCParserCommand>)receiver
{
	self = [super init];
	
	if (self)
	{
		_receiver = receiver;
	}
	
	return self;
}



/*
** TCParser - Parsing
*/
#pragma mark - TCParser - Parsing

- (void)parseLine:(NSData *)line
{
	if (!line)
		return;
	
	// Unscape protocol special chars.
	NSMutableData *mutableLine = [line mutableCopy];
	
	[mutableLine replaceCStr:"\\n" withCStr:"\n"];
	[mutableLine replaceCStr:"\\/" withCStr:"\\"];
	
	// Eplode the line from spaces.
	NSArray *items = [mutableLine explodeWithMaxFields:1 withFieldSeparator:" "];
	
	if ([items count] == 0)
		return;
	
	[self parseCommand:items];
}

- (void)parseCommand:(NSArray *)items
{
	if ([items count] == 0)
        return;
    
    NSString	*command = [[NSString alloc] initWithData:items[0] encoding:NSASCIIStringEncoding];
	NSData		*parameters = nil;
	
	if ([items count] > 1)
		parameters = items[1];
		
    // Dispatch command
    if ([command isEqualToString:@"ping"])
		[self parsePing:parameters];
    else if ([command isEqualToString:@"pong"])
        [self parsePong:parameters];
    else if ([command isEqualToString:@"status"])
        [self parseStatus:parameters];
    else if ([command isEqualToString:@"version"])
        [self parseVersion:parameters];
	else if ([command isEqualToString:@"client"])
        [self parseClient:parameters];
	else if ([command isEqualToString:@"profile_name"])
        [self parseProfileName:parameters];
	else if ([command isEqualToString:@"profile_text"])
        [self parseProfileText:parameters];
	else if ([command isEqualToString:@"profile_avatar_alpha"])
		 [self parseProfileAvatarAlpha:parameters];
	else if ([command isEqualToString:@"profile_avatar"])
        [self parseProfileAvatar:parameters];
	else if ([command isEqualToString:@"message"])
        [self parseMessage:parameters];
	else if ([command isEqualToString:@"add_me"])
        [self parseAddMe:parameters];
	else if ([command isEqualToString:@"remove_me"])
        [self parseRemoveMe:parameters];
	else if ([command isEqualToString:@"filename"])
        [self parseFileName:parameters];
	else if ([command isEqualToString:@"filedata"])
        [self parseFileData:parameters];
	else if ([command isEqualToString:@"filedata_ok"])
        [self parseFileDataOk:parameters];
	else if ([command isEqualToString:@"filedata_error"])
        [self parseFileDataError:parameters];
	else if ([command isEqualToString:@"file_stop_sending"])
        [self parseFileStopSending:parameters];
	else if ([command isEqualToString:@"file_stop_receiving"])
        [self parseFileStopReceiving:parameters];
    else
	{
		NSString *error = [NSString stringWithFormat:@"Unknown command '%@'", command];

		[self parserError:tcrec_unknown_command withString:error];
	}
}


- (void)parsePing:(NSData *)parameters
{
	NSArray *args = [parameters explodeWithMaxFields:2 withFieldSeparator:" "];
	
	// Check args.
	if ([args count] != 2)
    {
		[self parserError:tcrec_cmd_ping withString:@"Bad ping argument"];
        return;
    }
	
	// Parse command.
	NSString *address = [[NSString alloc] initWithData:args[0] encoding:NSASCIIStringEncoding];
	NSString *random = [[NSString alloc] initWithData:args[1] encoding:NSASCIIStringEncoding];
	
	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parser:parsedPingWithAddress:random:)])
		[receiver parser:self parsedPingWithAddress:address random:random];
	else
		[self parserError:tcrec_cmd_ping withString:@"Ping: Not handled"];
}

- (void)parsePong:(NSData *)parameters
{
	// Check args.
	if ([parameters length] == 0)
    {
		[self parserError:tcrec_cmd_pong withString:@"Bad pong argument"];
        return;
	}
	
	// Parse command.
	NSString *random = [[NSString alloc] initWithData:parameters encoding:NSASCIIStringEncoding];
	
	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parser:parsedPongWithRandom:)])
		[receiver parser:self parsedPongWithRandom:random];
	else
		[self parserError:tcrec_cmd_pong withString:@"Pong: Not handled"];
}

- (void)parseStatus:(NSData *)parameters
{
	// Check args.
	if ([parameters length] == 0)
    {
		[self parserError:tcrec_cmd_status withString:@"Bad status argument"];
        return;
	}
	
	// Parse command.
	NSString *status = [[NSString alloc] initWithData:parameters encoding:NSASCIIStringEncoding];
	
	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parser:parsedStatus:)])
		[receiver parser:self parsedStatus:status];
	else
		[self parserError:tcrec_cmd_status withString:@"Status: Not handled"];
}

- (void)parseVersion:(NSData *)parameters
{
	// Check args.
	if ([parameters length] == 0)
    {
		[self parserError:tcrec_cmd_version withString:@"Bad version argument"];
        return;
	}
	
	// Parse command.
	NSString *version = [[NSString alloc] initWithData:parameters encoding:NSASCIIStringEncoding];
	
	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parser:parsedVersion:)])
		[receiver parser:self parsedVersion:version];
	else
		[self parserError:tcrec_cmd_status withString:@"Version: Not handled"];
}

- (void)parseClient:(NSData *)parameters
{
	// Check args.
	if ([parameters length] == 0)
	{
		[self parserError:tcrec_cmd_client withString:@"Empty client argument"];
        return;
	}

	// Parse command.
	NSString *client = [[NSString alloc] initWithData:parameters encoding:NSUTF8StringEncoding];
	
	if (!client)
		return;
	
	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parser:parsedClient:)])
		[receiver parser:self parsedClient:client];
	else
		[self parserError:tcrec_cmd_client withString:@"Client: Not handled"];
}

- (void)parseProfileText:(NSData *)parameters
{
	// Parse command.
	NSString *text = @"";
	
	if (parameters)
		text = [[NSString alloc] initWithData:parameters encoding:NSUTF8StringEncoding];
	
	if (!text)
		return;
	
	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parser:parsedProfileText:)])
		[receiver parser:self parsedProfileText:text];
	else
		[self parserError:tcrec_cmd_profile_text withString:@"Profile-Text: Not handled"];
}

- (void)parseProfileName:(NSData *)parameters
{
	// Parse command.
	NSString *name = @"";
	
	if (parameters)
		name = [[NSString alloc] initWithData:parameters encoding:NSUTF8StringEncoding];
	
	if (!name)
		return;
	
	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parser:parsedProfileName:)])
		[receiver parser:self parsedProfileName:name];
	else
		[self parserError:tcrec_cmd_profile_name withString:@"Profile-Name: Not handled"];
}

- (void)parseProfileAvatar:(NSData *)parameters
{
	// Check command.
	if (!parameters)
		return;
	
	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parser:parsedProfileAvatar:)])
		[receiver parser:self parsedProfileAvatar:parameters];
	else
		[self parserError:tcrec_cmd_profile_avatar withString:@"Profile-Avatar: Not handled"];
}

- (void)parseProfileAvatarAlpha:(NSData *)parameters
{
	// Check command.
	if (!parameters)
		return;
	
	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parser:parsedProfileAvatarAlpha:)])
		[receiver parser:self parsedProfileAvatarAlpha:parameters];
	else
		[self parserError:tcrec_cmd_profile_avatar_alpha withString:@"Profile-AvatarAlpha: Not handled"];
}

- (void)parseMessage:(NSData *)parameters
{
	// Check args.
	if ([parameters length] == 0)
	{
		[self parserError:tcrec_cmd_message withString:@"Empty message content"];
        return;
	}
	
	// Parse command.
	NSString *message = [[NSString alloc] initWithData:parameters encoding:NSUTF8StringEncoding];
	
	if (!message)
		return;
	
	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parser:parsedMessage:)])
		[receiver parser:self parsedMessage:message];
	else
		[self parserError:tcrec_cmd_message withString:@"Message: Not handled"];
}

- (void)parseAddMe:(NSData *)parameters
{
	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parserParsedAddMe:)])
		[receiver parserParsedAddMe:self];
	else
		[self parserError:tcrec_cmd_addme withString:@"AddMe: Not handled"];
}

- (void)parseRemoveMe:(NSData *)parameters
{
	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parserparsedRemoveMe:)])
		[receiver parserparsedRemoveMe:self];
	else
		[self parserError:tcrec_cmd_removeme withString:@"RemoveMe: Not handled"];
}

- (void)parseFileName:(NSData *)parameters
{
	// Check args.
	NSArray *args = [parameters explodeWithMaxFields:3 withFieldSeparator:" "];
	
	if ([args count] != 4)
    {
		[self parserError:tcrec_cmd_filename withString:@"Bad filename argument"];
        return;
	}
	
	// Parse command.
	NSString *uuid = [[NSString alloc] initWithData:args[0] encoding:NSASCIIStringEncoding];
	NSString *fileSize = [[NSString alloc] initWithData:args[1] encoding:NSASCIIStringEncoding];
	NSString *blockSize = [[NSString alloc] initWithData:args[2] encoding:NSASCIIStringEncoding];
	NSString *fileName = [[NSString alloc] initWithData:args[3] encoding:NSUTF8StringEncoding];

	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parser:parsedFileNameWithUUIDD:fileSize:blockSize:fileName:)])
		[receiver parser:self parsedFileNameWithUUIDD:uuid fileSize:fileSize blockSize:blockSize fileName:fileName];
	else
		[self parserError:tcrec_cmd_filename withString:@"FileName: Not handled"];
}

- (void)parseFileData:(NSData *)parameters
{
	// Check args.
	NSArray *args = [parameters explodeWithMaxFields:3 withFieldSeparator:" "];
	
	if ([args count] != 4)
    {
		[self parserError:tcrec_cmd_filedata withString:@"Bad filedata argument"];
        return;
	}
	
	// Parse command.
	NSString	*uuid = [[NSString alloc] initWithData:args[0] encoding:NSASCIIStringEncoding];
	NSString	*start = [[NSString alloc] initWithData:args[1] encoding:NSASCIIStringEncoding];
	NSString	*hash = [[NSString alloc] initWithData:args[2] encoding:NSASCIIStringEncoding];
	NSData		*fileData = args[3];

	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parser:parsedFileDataWithUUID:start:hash:data:)])
		[receiver parser:self parsedFileDataWithUUID:uuid start:start hash:hash data:fileData];
	else
		[self parserError:tcrec_cmd_filedata withString:@"FileData: Not handled"];
}

- (void)parseFileDataOk:(NSData *)parameters
{
	// Check args.
	NSArray *args = [parameters explodeWithMaxFields:2 withFieldSeparator:" "];
	
	if ([args count] != 2)
    {
		[self parserError:tcrec_cmd_filedataok withString:@"Bad filedataok argument"];
        return;
	}
	
	// Parse command.
	NSString	*uuid = [[NSString alloc] initWithData:args[0] encoding:NSASCIIStringEncoding];
	NSString	*start = [[NSString alloc] initWithData:args[1] encoding:NSASCIIStringEncoding];
	
	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parser:parsedFileDataOkWithUUID:start:)])
		[receiver parser:self parsedFileDataOkWithUUID:uuid start:start];
	else
		[self parserError:tcrec_cmd_filedataok withString:@"FileDataOk: Not handled"];
}

- (void)parseFileDataError:(NSData *)parameters
{
	// Check args.
	NSArray *args = [parameters explodeWithMaxFields:2 withFieldSeparator:" "];

	if ([args count] != 2)
    {
		[self parserError:tcrec_cmd_filedataerror withString:@"Bad filedataerror argument"];
        return;
	}
	
	// Parse command.
	NSString	*uuid = [[NSString alloc] initWithData:args[0] encoding:NSASCIIStringEncoding];
	NSString	*start = [[NSString alloc] initWithData:args[1] encoding:NSASCIIStringEncoding];
	
	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parser:parsedFileDataErrorWithUUID:start:)])
		[receiver parser:self parsedFileDataErrorWithUUID:uuid start:start];
	else
		[self parserError:tcrec_cmd_filedataerror withString:@"FileDataError: Not handled"];
}

- (void)parseFileStopSending:(NSData *)parameters
{
	// Check args.
	if ([parameters length] == 0)
    {
		[self parserError:tcrec_cmd_filestopsending withString:@"Bad filestopsending argument"];
        return;
	}
	
	// Parse command.
	NSString *uuid = [[NSString alloc] initWithData:parameters encoding:NSASCIIStringEncoding];
	
	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parser:parsedFileStopSendingWithUUID:)])
		[receiver parser:self parsedFileStopSendingWithUUID:uuid];
	else
		[self parserError:tcrec_cmd_filestopsending withString:@"FileStopSending: Not handled"];
}

- (void)parseFileStopReceiving:(NSData *)parameters
{
	// Check args.
	if ([parameters length] == 0)
    {
		[self parserError:tcrec_cmd_filestopreceiving withString:@"Bad filestopreceiving argument"];
        return;
	}
	
	// Parse command.
	NSString *uuid = [[NSString alloc] initWithData:parameters encoding:NSASCIIStringEncoding];
	
	// Give to receiver.
	id <TCParserCommand> receiver = _receiver;
	
	if ([receiver respondsToSelector:@selector(parser:parsedFileStopReceivingWithUUID:)])
		[receiver parser:self parsedFileStopReceivingWithUUID:uuid];
	else
		[self parserError:tcrec_cmd_filestopreceiving withString:@"FileStopReceiving: Not handled"];
}



/*
** TCParser - Error
*/
#pragma mark - TCParser - Error

- (void)parserError:(tcrec_error)errorCode withString:(NSString *)string
{
	id <TCParserDelegate> delegate = _delegate;
	
	[delegate parser:self errorWithCode:errorCode andInformation:string];
}

@end
